Sources/PackageCollectionsSigning/Certificate/CertificatePolicy.swift

```diff
@@ -82,19 +95,333 @@ extension CertificatePolicy {
+            return wrappedCallback(.failure(CertificatePolicyError.ocspSetupFailure))
+        }
+
+        let requestData = Data(UnsafeBufferPointer(start: out, count: count))
+
+        let results = ThreadSafeArrayStore<Result<Bool, Error>>()
+        let group = DispatchGroup()
+
+        // Query each OCSP responder and record result
+        for index in 0 ..< ocspURLCount {
+            guard let urlStr = CCryptoBoringSSL_sk_OPENSSL_STRING_value(ocspURLs, numericCast(index)),
+                let url = String(validatingUTF8: urlStr).flatMap({ URL(string: $0) }) else {
+                results.append(.failure(OCSPError.badURL))
+                continue
+            }
+
+            let cacheKey = CacheKey(url: url, request: requestData)
+            if let cachedResult = self.resultCache[cacheKey] {
+                if cachedResult.timestamp + self.cacheTTL > DispatchTime.now() {
+                    results.append(.success(cachedResult.isCertGood))
+                    continue
+                }
+            }
+
+            var headers = HTTPClientHeaders()
+            headers.add(name: "Content-Type", value: "application/ocsp-request")
+            guard let host = url.host else {
+                results.append(.failure(OCSPError.badURL))
+                continue
+            }
+            headers.add(name: "Host", value: host)
+
+            var options = HTTPClientRequest.Options()
+            options.validResponseCodes = [200]
+
+            group.enter()
+            httpClient.post(url, body: requestData, headers: headers, options: options) { result in
+                defer { group.leave() }
+
+                switch result {
+                case .failure(let error):
+                    results.append(.failure(error))
+                case .success(let response):
+                    guard let responseData = response.body else {
+                        results.append(.failure(OCSPError.emptyResponseBody))
+                        return
+                    }
+
+                    let bytes = responseData.copyBytes()
+
+                    // Convert response to bio then OCSP response
+                    let bio = CCryptoBoringSSL_BIO_new(CCryptoBoringSSL_BIO_s_mem())
+                    defer { CCryptoBoringSSL_BIO_free(bio) }
+                    guard CCryptoBoringSSL_BIO_write(bio, bytes, numericCast(bytes.count)) > 0 else {
+                        results.append(.failure(OCSPError.responseConversionFailure))
+                        return
+                    }
+
+                    let response = d2i_OCSP_RESPONSE_bio(bio, nil)
+                    defer { OCSP_RESPONSE_free(response) }
+                    let basicResp = OCSP_response_get1_basic(response)
+                    defer { OCSP_BASICRESP_free(basicResp) }
+
+                    // This is just the OCSP response status, not the certificate's status
+                    guard OCSP_response_status(response) == OCSP_RESPONSE_STATUS_SUCCESSFUL,
```

Comment:
I only just remembered to double check this, and I'm glad I did: this doesn't actually validate the OCSP response. The OCSP response is digitally signed, and you need to verify that digital signature before you trust it. This logic is implemented by `OCSP_basic_verify` and friends from OpenSSL upstream.

Reply:
Thanks for checking this. I added `OCSP_basic_verify` in https://github.com/apple/swift-package-manager/pull/3272/commits/ed1614308e6fef91c4ab524ecb2f42de37d228f7